### PR TITLE
Adds train script

### PR DESCRIPTION
**Pull request by morganmcg1**
Based on HF flax language modelling examples: https://github.com/huggingface/transformers/blob/master/examples/flax/language-modeling/run_mlm_flax.py

This adds:
- a fix for `fixed_pos_embedding` to use jax's version of einsum
- a first start of a train script, steps though training steps on Google colab
    - The train script uses `FlaxRotoBARTForConditionalGeneration`  and a dummy `DummyFlaxDataCollatorForRotoBARTMLM`
- Dataset streaming from The Pile 
- Improved batch iteration, based on [HF dataset streaming] example script (https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py)
 

## To Do
Lots to do:
- ~~Get it to work on TPU~~
- - ~~Look closer at `generate_batch_splits`~~
- ~~Look closer at `group_texts`~~
- add sentence permutation 
- add text infill 
- Test on our TPU VM 

Test with:
```
python rotobart/run_dnlm_flax.py \
  --output_dir . \
  --overwrite_output_dir \
  --model_name_or_path rotobart \
  --tokenizer_name facebook/bart-large-cnn \
  --shuffle_buffer_size 100_000 \
  --do_train --do_eval \
  --max_seq_length 1024 \
  --encoder_layers 2 \
  --decoder_layers 2 \
  --per_device_train_batch_size 2 \
  --per_device_eval_batch_size 2 \
  --num_train_steps 1000 \
  --logging_steps 8 \
  --eval_steps 1000 \
  --num_eval_samples 100 \
  --warmup_steps 30 \
  --learning_rate 1e-4 \
  --use_wandb \
  --testing --colab_tpu
```

### CODE REVIEW

**Review by LouisCastricato — APPROVED** [review]

File: run_dnlm_flax.py [hunk 1]
```diff
@@ -0,0 +1,1387 @@
...
+        rng, input_rng = jax.random.split(rng)
+
+        # Generate an epoch by shuffling sampling indices from the train dataset
+        num_train_samples = len(tokenized_datasets["train"])
+        train_samples_idx = jax.random.permutation(input_rng, jnp.arange(num_train_samples))
+        train_batch_idx = generate_batch_splits(train_samples_idx, train_batch_size)
+
+        # Gather the indexes for creating the batch and do a training step
+        for i, batch_idx in enumerate(tqdm(train_batch_idx, desc="Training...", position=1)):
+            samples = [tokenized_datasets["train"][int(idx)] for idx in batch_idx]
```
Comment (mattbui) [inline run_dnlm_flax.py]: For streaming dataset, it'll be an iterable dataset so you cannot query the items via index like this, for shuffling the dataset, take a look at this: https://huggingface.co/docs/datasets/master/dataset_streaming.html#shuffling-the-dataset-shuffle

File: run_dnlm_flax.py [hunk 2]
```diff
@@ -0,0 +1,1387 @@
...
+            model_inputs = data_collator(samples)
```
Comment (mattbui) [inline run_dnlm_flax.py]: How about using the `data_collator` as a mapping function for the datasets:
```python
tokenized_datasets.map(data_collator, ...)
```

**Review by mattbui — COMMENTED** [review]
If we're planning to use a streaming dataset, I have some comments

File: run_dnlm_flax.py (hunk 2 above)
Comment (morganmcg1) [reply run_dnlm_flax.py]: To do the data collation ahead of time, instead of on the fly?

Comment (mattbui) [reply run_dnlm_flax.py]: @morganmcg1 no, it will apply the map functions on the fly. Here's the data flow I'd suggest:

```python
dataset = load_dataset(...)
dataset = dataset.map(permutation) # adding permutation text
dataset = dataset.map(tokenization) # tokenizeing text and permutation text to input_ids
dataset = dataset.map(text_filling)

for batch in dataloader(dataset):
	# training loop
```

File: run_dnlm_flax.py (hunk 1 above)
Comment (morganmcg1) [reply run_dnlm_flax.py]: Added commit to iterate through the batches using the [HF dataset streaming](https://github.com/huggingface/transformers/blob/master/examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py) example

File: run_dnlm_flax.py (hunk 2 above)
Comment (morganmcg1) [reply run_dnlm_flax.py]: Gotcha, looks good!

**Comment by LouisCastricato** [comment]
Adding text infilling now
